File: 059_Merge Two Sorted Lists.py
"""
Merge Two Sorted Lists

Merge two sorted linked lists and return it as a new sorted list.
"""

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def merge_two_lists(l1: ListNode, l2: ListNode) -> ListNode:
    dummy = ListNode(0)
    tail = dummy
    while l1 and l2:
        if l1.val < l2.val:
            tail.next = l1; l1 = l1.next
        else:
            tail.next = l2; l2 = l2.next
        tail = tail.next
    tail.next = l1 or l2
    return dummy.next

def print_list(node: ListNode):
    vals = []
    while node:
        vals.append(str(node.val)); node = node.next
    print(' '.join(vals))


if __name__ == '__main__':
    a = ListNode(1); a.next = ListNode(2); a.next.next = ListNode(4)
    b = ListNode(1); b.next = ListNode(3); b.next.next = ListNode(4)
    r = merge_two_lists(a,b)
    print('Merged:', end=' '); print_list(r)

# Time Complexity: O(m+n)
# Space Complexity: O(1)

File: 059_Merge Two Sorted Lists.cpp
/*
Merge Two Sorted Lists

Merge two sorted linked lists and return it as a new sorted list.
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int val; ListNode* next;
    ListNode(int x): val(x), next(nullptr) {}
};

ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
    ListNode dummy(0);
    ListNode* tail = &dummy;
    while (l1 && l2) {
        if (l1->val < l2->val) { tail->next = l1; l1 = l1->next; }
        else { tail->next = l2; l2 = l2->next; }
        tail = tail->next;
    }
    tail->next = l1 ? l1 : l2;
    return dummy.next;
}

void printList(ListNode* head) { while (head) { cout << head->val; if (head->next) cout << ' '; head = head->next; } cout << endl; }

int main() {
    ListNode* a = new ListNode(1); a->next = new ListNode(2); a->next->next = new ListNode(4);
    ListNode* b = new ListNode(1); b->next = new ListNode(3); b->next->next = new ListNode(4);
    auto r = mergeTwoLists(a,b);
    cout << "Merged: "; printList(r); // 1 1 2 3 4 4
    return 0;
}

// Time Complexity: O(m+n)
// Space Complexity: O(1)
